explorer-backend/fly/utils/coin_gecko.go

```diff
@@ -86,52 +76,126 @@ func rangeFromTime(t time.Time, hours int) (start time.Time, end time.Time) {
 	return t.Add(-duration), t.Add(duration)
 }
 
-func FetchCoinGeckoPrice(coinId string, timestamp time.Time) (float64, error) {
-	start, end := rangeFromTime(timestamp, 12)
-	priceUrl := fmt.Sprintf("%vcoins/%v/market_chart/range?vs_currency=usd&from=%v&to=%v", coinGeckoUrl, coinId, start.Unix(), end.Unix())
-	req, reqErr := http.NewRequest("GET", priceUrl, nil)
-	if reqErr != nil {
-		return 0, reqErr
-	}
-
-	res, resErr := http.DefaultClient.Do(req)
+func fetch[T any](request *http.Request) (*T, error) {
+	res, resErr := http.DefaultClient.Do(request)
 	if resErr != nil {
-		return 0, resErr
+		return nil, fmt.Errorf("failed to fetch, err: %v", resErr)
 	}
 	if res.StatusCode >= 400 {
-		return 0, fmt.Errorf("failed to get CoinGecko prices")
+		return nil, fmt.Errorf("invalid status code: %v", res.StatusCode)
 	}
 
 	defer res.Body.Close()
 	body, bodyErr := ioutil.ReadAll(res.Body)
 	if bodyErr != nil {
-		return 0, bodyErr
+		return nil, fmt.Errorf("failed to read response body, err: %v", bodyErr)
 	}
 
-	var parsed CoinGeckoMarketRes
-	parseErr := json.Unmarshal(body, &parsed)
+	var result T
+	parseErr := json.Unmarshal(body, &result)
 	if parseErr != nil {
 		var errRes CoinGeckoErrorRes
 		if err := json.Unmarshal(body, &errRes); err == nil {
-			return 0, fmt.Errorf(errRes.Error)
+			return nil, fmt.Errorf(errRes.Error)
 		}
-		return 0, parseErr
+		return nil, fmt.Errorf("failed to decode response, err: %v", parseErr)
+	}
+	return &result, nil
+}
+
+func FetchCoinGeckoPrice(coinId string, timestamp time.Time) (float64, error) {
+	start, end := rangeFromTime(timestamp, 12)
+	priceUrl := fmt.Sprintf("%vcoins/%v/market_chart/range?vs_currency=usd&from=%v&to=%v", coinGeckoUrl, coinId, start.Unix(), end.Unix())
+	req, reqErr := http.NewRequest("GET", priceUrl, nil)
+	if reqErr != nil {
+		return 0, reqErr
+	}
+
+	result, err := fetch[CoinGeckoMarketRes](req)
+	if err != nil {
+		return 0, err
 	}
-	if len(parsed.Prices) >= 1 {
+	if len(result.Prices) >= 1 {
 		hourAgo := time.Now().Add(-time.Duration(1) * time.Hour)
 		withinLastHour := timestamp.After(hourAgo)
 		var priceIndex int
 		if withinLastHour {
 			// use the last price in the list, latest price
-			priceIndex = len(parsed.Prices) - 1
+			priceIndex = len(result.Prices) - 1
 		} else {
 			// use a price from the middle of the list, as that should be
 			// closest to the timestamp.
-			numPrices := len(parsed.Prices)
+			numPrices := len(result.Prices)
 			priceIndex = numPrices / 2
 		}
-		price := parsed.Prices[priceIndex][1]
+		price := result.Prices[priceIndex][1]
 		return price, nil
 	}
 	return 0, fmt.Errorf("no price found for %v", coinId)
 }
+
+type Price struct {
+	USD float64 `json:"usd"`
+}
+type CoinGeckoCoinPrices map[string]Price
+
+func fetchCoinGeckoPrices(coinIds []string) (map[string]float64, error) {
```

Comment:
Why does the backend needs to fetch token price? It seems to be the work of front-end

Reply:
The code is migrated from the old wormhole explorer-backend. The purpose of fetching the token price is for statistics. When the explorer-backend receives a new VAA, it will get the token price and then update the statistics.